src/common/type/collection.h

```diff
@@ -0,0 +1,163 @@
+/***********************************************************************************************************************************
+Define the interface for an abstract, iterable Collection and provide "syntactic sugar" for scanning across collections which
+implement the "newIter()" method. For information on how to create an "Iterable" collection, see the "Iterator.c" file.
+These interfaces are inspired by Rust's Collection and Iterable traits.
+
+Using "syntactic sugar" to scan a list,
+    FOREACH(ItemType, item, List, list)
+        doSomething(*item);
+    ENDFOREACH;
+
+This file also defines an abstract Collection type, which is a wrapper around any iterable collection.
+Once wrapped, the Collection can be passed around and the users don't need to know the details of what type
+of collection is inside.
+
+    // Construct an abstract Collection which wraps the List.
+    Collection *collection = NEWCOLLECTION(List, list);
+
+    // Iterate through the abstract Collection just like any other collection.
+    FOREACH(ItemType, item, Collection, collection)
+        doSomething(*item)
+    ENDFOREACH;
+
+***********************************************************************************************************************************/
+#ifndef PGBACKREST_COLLECTION_H
+#define PGBACKREST_COLLECTION_H
+
+#include <stdbool.h>
+#include "common/assert.h"
+
+// TODO: Temporary. There must be a reason this hasn't been done. Perhaps test mocks need the "attribute()" syntax visible.
```

Comment:
If "never got around to it" is a good reason, then yes. Here's a PR that creates a macro for this commonly used pattern: https://github.com/pgbackrest/pgbackrest/pull/1861.